activesupport/lib/active_support/core_ext/range/intersection.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class Range
+  # The maximum sized range covered by both self AND other
+  # Returns nil if there is no overlap between self and other
+  def intersection(other)
+    unless other.is_a?(::Range)
+      raise TypeError, "Can't intersect a range by a(n) #{other.class}"
+    end
+
+    if (inverted_range = [self, other].find { |r| r.inverted? })
```

Comment:
could you walk me through the reasoning as to why we can't intersect inverted ranges? What is your definition of intersection here?

Reply:
thanks for the thorough explanation, truly appreciated.

Yeah I can see the ambiguity that arises here. One option might be to raise an error like you've done, another option might be to retain the structure of the left hand side, what are your thoughts on that?

Also, not sure if you saw my reply on your rails thread, but you might be interested in https://github.com/hernanat/extended_range

Would welcome the PR for this over there, and we can discuss further if you're interested

Reply:
To be clear my perspective is basically that on an element-by-element basis, the range `5..1` contains (individual) elements of `2..3`, and so it might be reasonable to keep the structure of the LHS if we're considering it like that.

Reply:
> A `Range` IS a set...it doesn't descend from Set explicitly, but it "represents an interval—a set" (quote taken from ruby doc.

I take your point, to be fair though I think the docs are using a very loose definition of "set" here ... one of the things that bothers me about Ruby's `Range` class is the fact that it does _not_ respect the mathematical definition of an interval / set, and so the ambiguities such as the one we've discussed here come up 😞

Reply:
Thanks @hernanat! I'll take a look at your library and consider a PR over there (along with your suggestion for retaining the structure (order?) of the LHS rather than raising...interesting, though I'd want to maintain some consistency with the def of `include?`).

Reply:
of course! also your work has made me rethink some decisions that I've made over there, and so I appreciate you for that!